homeworks/solutions/krasnor/5-remesh/src/halfedge.js

```diff
@@ -0,0 +1,821 @@
+            real_face_cnt++;
+          }else{
+            deleted_face_cnt++;
+          }
+        }
+      );
+
+      console.log("in collapse: %s - current cnt faces %s || real face count: %s, deleted: %s", i, curr_face_cnt, real_face_cnt, deleted_face_cnt);
+      let curr_hf = e.halfedge;
+      let curr_hf_twin = curr_hf.twin;
+      let curr_p1 = curr_hf.vertex;
+      let curr_p2 = curr_hf_twin.vertex;
+
+      // create new vertex along edge
+      let new_vertex = new Vertex();
+      new_vertex.idx = nxt_idx_vertices++;
+      new_vertex.position = (new Vector()).add(m);
+      new_vertex.halfedge = curr_hf.prev.twin;
+
+      if(new_vertex.halfedge.idx === -1){
+        console.log("new_vertex.halfedge.idx is invalid");
+      }
+
+      // connect old halfedges to new vertex
+      curr_p1.halfedges(
+          (hf, i) => {
+            // console.log("overwriting hf: %s, point: %s with point: %s" , hf.idx, hf.vertex.idx, new_vertex.idx)
+            hf.vertex = new_vertex;
+          }
+      );
+      curr_p2.halfedges(
+          (hf, i) => {
+            // console.log("overwriting hf: %s, point: %s with point: %s" , hf.idx, hf.vertex.idx, new_vertex.idx)
+            hf.vertex = new_vertex;
+          }
+      );
+
+      // merge faces and their edges
+      [curr_hf, curr_hf_twin].forEach(
+          i_hf => {
+          let f = i_hf.face;
+          if(f != null){
+            //     c
+            //    / ^
+            // L / f \ R
+            //  v     \
+            //  a----->b
+            //   i_hf
+
+            let currFaceIdx = f.idx;
+            let l = i_hf.prev;
+            let r = i_hf.next;
+            let l_e = l.edge;
+            let r_e = r.edge;
+            let l_t = l.twin;
+            let r_t = r.twin;
+            let new_edge = new Edge();
+            new_edge.idx = nxt_idx_edges++;
+            new_edge.halfedge = l_t;
+
+
+            // update errors
+            let c = l.vertex;
+            c.old_id = c.idx;
+            c.halfedge = r_t;
+
+            l_t.twin = r_t;
+            r_t.twin = l_t;
+            l_t.edge.idx = -1;
+            r_t.edge.idx = -1;
+            l_t.edge = new_edge;
+            r_t.edge = new_edge;
+
+            // mark as obsolete
+            l.idx = -1;
+            r.idx = -1;
+            l_e.idx = -1;
+            r_e.idx = -1;
+            f.idx = -1;
+
+            this.edges.push(new_edge);
+            faces_removed++;
+          }
+        }
+      );
+
+
+      // mark current edge, halfedges and vertices as obsolete
+      curr_p1.idx = -1;
+      curr_p2.idx = -1;
+      e.idx = -1;
+      curr_hf.idx = -1;
+      curr_hf.twin.idx = -1;
+
+      // set new hf of vertex
+      this.vertices.push(new_vertex);
+      curr_face_cnt -= faces_removed;
+
+      // recalculation the whole heap as it contains outdated pairs/edges
+      heap = null;
+      heap = new PriorityQueue(
+          (a, b) => a.error < b.error // min top
```

Comment:
It would be much easier if you push edges directly into the priority queue, and compare them using an Edge.error() method